### PR TITLE
Add toxcore capabilities to "online" packet sent to friends

### PR DESCRIPTION
**Pull request by zoff99**
This change is [Reviewable](https://reviewable.io/reviews/toktok/c-toxcore/1033)

### CODE REVIEW

File: toxcore/Messenger.c [hunk 1]
```diff
@@ -152,16 +152,30 @@ void getaddress(const Messenger *m, uint8_t *address)
...
         return 0;
     }
 
+    uint8_t buf[TOX_CAPABILITIES_SIZE + 1];
+    buf[0] = PACKET_ID_ONLINE;
+    net_pack_u64(buf + 1, TOX_CAPABILITIES_CURRENT);
+
+    write_cryptpacket(m->net_crypto, friend_connection_crypt_connection_id(m->fr_c,
+                      m->friendlist[friendnumber].friendcon_id), buf, TOX_CAPABILITIES_SIZE + 1, 0);
+    // TODO: what to do if res == -1 ?
```
Comment (iphydf) [inline toxcore/Messenger.c]: Probably return -1.

File: toxcore/Messenger.c [hunk 2]
```diff
@@ -2157,6 +2173,23 @@ static int m_handle_status(void *object, int i, uint8_t status, void *userdata)
     return 0;
 }
 
+/* get capabilities of friend's toxcore
+ * return TOX_CAPABILITY_BASIC on any error
```
Comment (iphydf) [inline toxcore/Messenger.c]: That's a surprising error behaviour. If the friend doesn't exist, you say that they have basic capabilities. In fact they have no capabilities at all.

File: toxcore/Messenger.h [hunk 3]
```diff
@@ -65,6 +65,17 @@ typedef struct Messenger_Options {
...
 
+enum ToxCapabilitiesFlags {
+    TOX_CAPABILITY_BASIC = 1 << 0,
+    TOX_CAPABILITY_PGC = 1 << 1,
+    TOX_CAPABILITY_H264 = 1 << 2,
+    TOX_CAPABILITY_MESSAGE_V2 = 1 << 3,
+};
+
+#define TOX_CAPABILITIES_CURRENT (uint64_t)(TOX_CAPABILITY_BASIC)
+#define TOX_CAPABILITIES_SIZE sizeof(uint64_t)
```
Comment (iphydf) [inline toxcore/Messenger.h]: The size of a capabilities packet is 1 + sizeof(uint64_t).

File: toxcore/Messenger.c [hunk 4]
```diff
@@ -2169,7 +2202,14 @@ static int m_handle_packet(void *object, int i, const uint8_t *temp, uint16_t le
     uint32_t data_length = len - 1;
 
     if (m->friendlist[i].status != FRIEND_ONLINE) {
-        if (packet_id == PACKET_ID_ONLINE && len == 1) {
+        if (packet_id == PACKET_ID_ONLINE) {
+
+            if ((len >= 1) && (len == TOX_CAPABILITIES_SIZE)) {
```
Comment (iphydf) [inline toxcore/Messenger.c]: Remove superfluous ().

File: toxcore/Messenger.c [hunk 5]
```diff
@@ -2169,7 +2202,14 @@ static int m_handle_packet(void *object, int i, const uint8_t *temp, uint16_t le
...
+                uint64_t received_caps = TOX_CAPABILITY_BASIC;
```
Comment (iphydf) [inline toxcore/Messenger.c]: Dead store.

File: toxcore/Messenger.c [hunk 6]
```diff
@@ -2169,7 +2202,14 @@ static int m_handle_packet(void *object, int i, const uint8_t *temp, uint16_t le
     uint32_t data_length = len - 1;
 
     if (m->friendlist[i].status != FRIEND_ONLINE) {
-        if (packet_id == PACKET_ID_ONLINE && len == 1) {
+        if (packet_id == PACKET_ID_ONLINE) {
+
```
Comment (iphydf) [inline toxcore/Messenger.c]: Remove empty line.

Comment (hugbubby) [reply toxcore/Messenger.c]: The line is removed, read the line numbers, github is just weird.

Comment (hugbubby) [reply toxcore/Messenger.c]: Ohhhhhhhh that line

File: toxcore/Messenger.h (hunk 3 above)
Comment (zoff99) [reply toxcore/Messenger.h]: but the pure data size of the capabilties is sizeof(uint64_t)

**Comment by zugz** [comment]
What is the purpose of the TOX_CAPABILITY_BASIC flag? What does it mean 
if it is set to false?

I don't think we should actually reserve flags for anything yet where we 
can't specify exactly what it will mean, as is the case at least for the 
PGC flag.

Separately: perhaps it would make sense to reserve the first bit for 
future expansion - if it's set, it means "ignore this bitmap, and look 
instead at the successor method of specifying capabilities" (whatever 
that turns out to be).

**Comment by zoff99** [comment]
@iphydf 
error: ISO C restricts enumerator values to range of 'int'
how can i make the flags 64bit wide?

**Comment by iphydf** [comment]
@zoff99 there are open comments from @zugz.

**Pull request closed without merging**